### PR TITLE
Preserve custom arguments passed on launch and working directory

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

The `Arguments Passed on Launch` and the custom working directory is often used as short-lived configuration, especially when working on CLIs. When working on tuist, it's very annoying that on `tuist generate`, you lose your setup (e.g., when you are debugging a specific a project). This PR fixes that pain.

image
image

### How to test the changes locally 🧐

Set up custom working directory and arguments passed on launch -> they should be preserved on `tuist generate`.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint:fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistGenerator/Generator/ProjectDescriptorGenerator.swift [hunk 1]
```diff
@@ -81,6 +81,8 @@ final class ProjectDescriptorGenerator: ProjectDescriptorGenerating {
     ) throws -> ProjectDescriptor {
         logger.notice("Generating project \(project.name)")
 
+        let previousXcodeProj = try? XcodeProj(pathString: project.xcodeProjPath.pathString)
```
Comment (pepicrft) [inline Sources/TuistGenerator/Generator/ProjectDescriptorGenerator.swift:84]: I'd benchmark this one to get a sense of the performance penalty. If it's significant, I'd consider adding a generation option, something along the lines of `.preserveLaunchConfiguration`.

Comment (fortmarek) [reply Sources/TuistGenerator/Generator/ProjectDescriptorGenerator.swift:84]: I did a quick test and it takes milliseconds

Comment (pepicrft) [reply Sources/TuistGenerator/Generator/ProjectDescriptorGenerator.swift:84]: Good!

**Comment by kwridan** [comment]
Thanks for flagging this @fortmarek - typically to resolve this we duplicate the scheme in question and make it a user scheme which should be left unaffected by the generation process and is also git ignored. 

Making the generation preserve past scheme modifications in shared schemes sadly leads to inconsistency in the overall generated projects. Those are expected to be identical each run on all environments. 

For our workflow we check in generated projects and lint projects to ensure the checked in version matches the generated one, typically a re-generate ahead of pushing a PR should get everything back in sync. This change makes it such developers will be stuck in an inconsistent state the second they manually modify a scheme. 

If there's a strong desire to have this built into tuist I'd move it out of the core generation logic and place in a mapper externally.

**Comment by fortmarek** [comment]
This can't be really put into a mapper since we don't allow to specify a custom working directory but we can make this opt-in via a generation option instead.

I reverted the PR for the time being before we add that option.

**Comment by kwridan** [comment]
Thanks for reverting 🙏

We can look at how to make mappers work for this, that said before re-introducing this with an option worth considering if a code change is needed at all - as mentioned there already exists a solution to the issue you were facing via leveraging user manually duplicated user schemes. 

Tuist already preserves those 

https://github.com/tuist/tuist/blob/main/Tests/TuistGeneratorIntegrationTests/Generator/XcodeProjWriterTests.swift#L140

**Pull request merged**